PIPS/pip-1.md

```diff
@@ -0,0 +1,147 @@
+---
+pip: 1
+title: Permissionless Duo LockProxy
+status: Draft
+author: John Wong (@jwheur)
+created: 2020-06-23
+---
+
+## Simple Summary
+The current LockProxy implementation requires an operator to control `setManagerProxy`, `bindProxyHash` and `bindAssetHash`.
+
+This PIP proposes an alternative implementation which does not require an operator.
+The tradeoff of this implementation is that an asset can be transferred between at most two chains.
+
+:white_check_mark: Supported Example:
+1. Lock token A => Ethereum LockProxy --> Neo LockProxy => Unlock token A
+2. Lock token A => Neo LockProxy --> Ethereum LockProxy => Unlock token A
+
+:x: Not Supported Example:
+1. Lock token A => Ethereum LockProxy --> Neo LockProxy => Unlock token A
+2. Lock token A => Neo LockProxy --> Ontology LockProxy => Unlock token A
+3. Lock token A => Ontology LockProxy --> Ethereum LockProxy => This will fail
+
+## Specification
+To support non-native tokens within the LockProxy, a token contract should be deployed to represent the non-native token.
+
+The token's constructor should call `delegateAsset` on the LockProxy that it is delegated to, for example:
+
+```
+constructor (address lockProxyContractAddress, uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash) public ERC20Detailed("ONT Token", "ONTX", 0) {
+    uint256 totalSupply = 1000000000;
+    _mint(lockProxyContractAddress, totalSupply);
+    lockProxy.delegateAsset(nativeChainId, nativeLockProxy, nativeAssetHash, totalSupply);
+}
+```
+
+In the LockProxy, `delegateAsset` can be implemented as:
+```
+struct DelegatedAsset {
+    uint64 nativeChainId;
+    bytes nativeLockProxy;
+    bytes nativeAssetHash;
+}
+
+mapping(address => DelegatedAsset) delegatedAssets;
+mapping(bytes32 => uint256) balances;
+
+function delegateAsset(uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash, uint256 totalSupply) public {
```

Comment:
Within this method, I think, we need to check the lockproxy contract account balance of assetHash=_msgSender() asset actually equal totalSupply. But that may not be mandatory.

Reply:
good point, I'll add that check, it will help make the process more robust